Make setting a link@rel=stylesheet to “disabled” remove the stylesheet

sideshowbarker opened the pull request:
#### 123f48fb7964cb8277f68beabb56ffbe5e87ee98
<pre>
Make setting a link@rel=stylesheet to “disabled” remove the stylesheet
<a href="https://bugs.webkit.org/show_bug.cgi?id=260642">https://bugs.webkit.org/show_bug.cgi?id=260642</a>

Reviewed by Chris Dumez.

When the “disabled” attribute is added to a &lt;link&gt; element with a
rel=stylesheet, this change causes the stylesheet to be removed, per the
requirements in <a href="https://drafts.csswg.org/cssom/#remove-a-css-style-sheet">https://drafts.csswg.org/cssom/#remove-a-css-style-sheet</a>
(which include setting the ownerNode to null) — making the behavior in
this case interoperable with existing behavior in Gecko and Blink.

Otherwise, without this change, the stylesheet isn’t removed — which breaks
spec conformance, and breaks interoperability with Gecko and Blink.

* LayoutTests/fast/dom/HTMLLinkElement/disabled-attribute-expected.txt:
* LayoutTests/fast/dom/HTMLLinkElement/disabled-attribute.html:
* LayoutTests/http/tests/css/link-css-disabled-value-with-slow-loading-sheet-expected.txt:
* LayoutTests/http/tests/css/link-css-disabled-value-with-slow-loading-sheet-in-error-expected.txt:
* LayoutTests/http/tests/css/link-css-disabled-value-with-slow-loading-sheet-in-error.html:
* LayoutTests/http/tests/css/link-css-disabled-value-with-slow-loading-sheet.html:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/HTMLLinkElement-disabled-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/HTMLLinkElement-disabled-002-expected.txt:
* LayoutTests/platform/mac-monterey-wk1/TestExpectations:
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::setDisabledState):

Canonical link: <a href="https://commits.webkit.org/269753@main">https://commits.webkit.org/269753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9065da88bd491d2a655d5d80fa957080da783fc3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23324 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24447 "Built successfully") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21546 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23595 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2994 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23880 "Built successfully") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23565 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1021 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20198 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26091 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/902 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21085 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27253 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21290 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21347 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25110 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/796 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18543 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/753 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20863 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5605 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1209 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1055 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (approved):
LGTM if the bots are happy.

sideshowbarker committed:
Make setting a link@rel=stylesheet to “disabled” remove the stylesheet
https://bugs.webkit.org/show_bug.cgi?id=260642

Reviewed by Chris Dumez.

When the “disabled” attribute is added to a <link> element with a
rel=stylesheet, this change causes the stylesheet to be removed, per the
requirements in https://drafts.csswg.org/cssom/#remove-a-css-style-sheet
(which include setting the ownerNode to null) — making the behavior in
this case interoperable with existing behavior in Gecko and Blink.

Otherwise, without this change, the stylesheet isn’t removed — which breaks
spec conformance, and breaks interoperability with Gecko and Blink.

* LayoutTests/fast/dom/HTMLLinkElement/disabled-attribute-expected.txt:
* LayoutTests/fast/dom/HTMLLinkElement/disabled-attribute.html:
* LayoutTests/http/tests/css/link-css-disabled-value-with-slow-loading-sheet-expected.txt:
* LayoutTests/http/tests/css/link-css-disabled-value-with-slow-loading-sheet-in-error-expected.txt:
* LayoutTests/http/tests/css/link-css-disabled-value-with-slow-loading-sheet-in-error.html:
* LayoutTests/http/tests/css/link-css-disabled-value-with-slow-loading-sheet.html:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/HTMLLinkElement-disabled-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/HTMLLinkElement-disabled-002-expected.txt:
* LayoutTests/platform/mac-monterey-wk1/TestExpectations:
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::setDisabledState):

Canonical link: https://commits.webkit.org/269753@main

fujii commented on LayoutTests/http/tests/css/link-css-disabled-value-with-slow-loading-sheet-in-error.html:
```diff
@@ -20,8 +20,8 @@
-shouldBeNonNull("mainSheetLink.sheet");
-shouldBeNull("alternateSheetLink.sheet");
+shouldBeNull("mainSheetLink.sheet");
+shouldBeNonNull("alternateSheetLink.sheet");
```
This test is randomly failing (even before your change).
https://results.webkit.org/?suite=layout-tests&test=http%2Ftests%2Fcss%2Flink-css-disabled-value-with-slow-loading-sheet.html
I think we can remove this line because alternateSheetLink.sheet is checked by the below code. WDTY?

sideshowbarker replied:
Alternatively, maybe we should remove the entire test from the repo sources?

That test is actually out of date with the current spec requirements — and looking through the repo history now, I see it’s a very old test (12+ years), and has been known for more than 9 years now to be flaky.

See https://github.com/WebKit/WebKit/commit/f2f23f38076cd3fad667a68544df4c8b595c9546, which notes both that the test doesn’t even seem to be testing any slowness it was intended to test against (_“These scripts were not "slow" at all”_) and that it also doesn’t align with how WebCore does stylesheet loading (_“WebCore has its own ideas about when to load stylesheets - notably, default stylesheet loading blocks script execution, so we cannot do anything before the stylesheet loads”_).

Further, as far as testing what’s required by the current relevant standards, it’s superseded by these WPTs:

- https://wpt.fyi/results/css/cssom/HTMLLinkElement-disabled-002.html
- https://wpt.fyi/results/css/cssom/HTMLLinkElement-disabled-001.html

…both of which Safari fails prior to TP181 — but the change in this PR, https://github.com/WebKit/WebKit/pull/19331, makes WebKit now pass.

Finally, per https://github.com/WebKit/WebKit/commit/c6a6d036f285fbaff5be9c601307ce9c5e71ca46 (https://bugs.webkit.org/show_bug.cgi?id=263679) the test was just today marked as `Failure` in the text expectations for yet another platform ([`mac-wk1/TestExpectations`](https://github.com/WebKit/WebKit/commit/c6a6d036f285fbaff5be9c601307ce9c5e71ca46#diff-a0da6d7c099dd9e2d22a140bbdf209f01f0f0394d19609d6d3e7c60a85ad760d)).

In summary it really appears that at this point it’s a test which is buying us nothing yet, continuing to cost us — and given that, it seems like we’d be better off just removing it.

So, should I open a PR to remove the test?
